lib/extras/gain_map.cc

```diff
@@ -0,0 +1,225 @@
+
+  ADVANCE_CURSOR(map_bundle->alt_icc_size);
+  memcpy(output_buffer + cursor, map_bundle->alt_icc, map_bundle->alt_icc_size);
+
+  ADVANCE_CURSOR(map_bundle->gain_map_size);
+  memcpy(output_buffer + cursor, map_bundle->gain_map,
+         map_bundle->gain_map_size);
+
+#undef ADVANCE_CURSOR
+
+  cursor = next_cursor;
+
+  if (bytes_written != nullptr)
+    *bytes_written = cursor;  // Ensure size_t compatibility
+  return cursor == output_buffer_size ? JXL_TRUE : JXL_FALSE;
+}
+
+JXL_BOOL JxlGainMapReadBundle(JxlGainMapBundle* map_bundle,
+                              const uint8_t* input_buffer,
+                              const size_t input_buffer_size,
+                              size_t* bytes_read) {
+  if (map_bundle == nullptr || input_buffer == nullptr ||
+      input_buffer_size == 0) {
+    return JXL_FALSE;
+  }
+
+  uint64_t cursor = 0;
+  uint64_t next_cursor = 0;
+
+#define ADVANCE_CURSOR(n)                        \
+  do {                                           \
+    cursor = next_cursor;                        \
+    if (!jxl::SafeAdd(cursor, n, next_cursor) || \
+        next_cursor > input_buffer_size) {       \
+      return JXL_FALSE;                          \
+    }                                            \
+  } while (false)
+
+  // Read the version byte
+  ADVANCE_CURSOR(1);
+  map_bundle->jhgm_version = input_buffer[cursor];
+
+  // Read and swap gain_map_metadata_size
+  ADVANCE_CURSOR(2);
+  uint16_t gain_map_metadata_size = LoadBE16(input_buffer + cursor);
+
+  ADVANCE_CURSOR(gain_map_metadata_size);
+  map_bundle->gain_map_metadata_size = gain_map_metadata_size;
+  map_bundle->gain_map_metadata = input_buffer + cursor;
+
+  // Read compressed_color_encoding_size
+  ADVANCE_CURSOR(1);
+  uint8_t compressed_color_encoding_size;
+  memcpy(&compressed_color_encoding_size, input_buffer + cursor, 1);
+
+  map_bundle->has_color_encoding = (compressed_color_encoding_size > 0);
+  if (map_bundle->has_color_encoding) {
+    ADVANCE_CURSOR(compressed_color_encoding_size);
+    // Decode color encoding
+    jxl::Span<const uint8_t> color_encoding_span(
+        input_buffer + cursor, compressed_color_encoding_size);
+    jxl::BitReader color_encoding_reader(color_encoding_span);
+    jxl::ColorEncoding internal_color_encoding;
+    if (!jxl::Bundle::Read(&color_encoding_reader, &internal_color_encoding)) {
+      return JXL_FALSE;
+    }
+    JXL_RETURN_IF_ERROR(color_encoding_reader.Close());
+    map_bundle->color_encoding = internal_color_encoding.ToExternal();
+  }
+
+  // Read and swap compressed_icc_size
+  ADVANCE_CURSOR(4);
+  uint32_t compressed_icc_size = LoadBE32(input_buffer + cursor);
+
+  ADVANCE_CURSOR(compressed_icc_size);
+  map_bundle->alt_icc_size = compressed_icc_size;
+  map_bundle->alt_icc = input_buffer + cursor;
+
+  // Calculate remaining bytes for gain map
+  cursor = next_cursor;
+  map_bundle->gain_map_size = input_buffer_size - cursor;
```

Comment:
Maybe add a comment that this is guaranteed not to underflow, and why

Reply:
done